plugins/modules/user.py

```diff
@@ -0,0 +1,261 @@
+  - Module updates selected virtual machine node affinity.
+version_added: 1.0.0
+extends_documentation_fragment:
+  - scale_computing.hypercore.cluster_instance
+seealso:
+  - module: scale_computing.hypercore.user_info
+options:
+  state:
+    description:
+      - The desired state of the snapshot schedule object.
+      - Use C(absent) to ensure the snapshot schedule will be absent from the API and C(present) to ensure
+        snapshot schedule will remain present on the API.
+    type: str
+    choices: [ present, absent ]
+    required: true
+  username:
+    description:
+      - The user name.
+      - Serves as unique identifier.
+      - If C(state) is I(present) and C(username) doesn't exist, a new user is created.
+      - If C(state) is I(present) and C(username) already exists, existing user is updated.
+      - If C(state) is I(absent) and C(username) is found, existing user is deleted.
+    type: str
+    required: True
+  new_username:
+    description:
+      - New user name.
+      - Relevant only if C(state) is I(present) and C(username) exists.
+    type: str
+  password:
+    description:
+      - Plain text password for authentication.
+    type: str
+  full_name:
+    description:
+      - Human-readable display name.
+    type: str
+  roles:
+    description:
+      - Names of the roles this user is a member of.
+    type: list
+    elements: str
+    choices: [ Backup, VM Delete, Cluster Settings, Cluster Shutdown, VM Power Controls, Read, VM Create/Edit, Admin ]
+  session_limit:
+    description:
+      - The maximum number of Sessions this user may have at one time.
+    type: int
+notes:
+  - C(check_mode) is not supported.
+"""
+
+EXAMPLES = r"""
+- name: Create new user
+  scale_computing.hypercore.user:
+    state: present
+    username: user
+    password: password
+    full_name: User
+    roles:
+      - VM delete
+      - VM Create/Edit
+      - VM Power Controls
+    session_limit: 0
+
+- name: Update existing user
+  scale_computing.hypercore.user:
+    state: present
+    username: user
+    new_username: new_username
+    password: new_password
+
+- name: Delete the user
+  scale_computing.hypercore.user:
+    state: absent
+    username: user
+"""
+
+RETURN = r"""
+record:
+  description:
+    - User record.
+  returned: success
+  type: dict
+  sample:
+    full_name: xlab
+    roles:
+      - Admin
+    session_limit: 0
+    username: xlab
+    uuid: 51e6d073-7566-4273-9196-58720117bd7f
+"""
+
+from ansible.module_utils.basic import AnsibleModule
+
+from ..module_utils import arguments, errors
+from ..module_utils.client import Client
+from ..module_utils.rest_client import RestClient
+from ..module_utils.rest_client import CachedRestClient
+from ..module_utils.user import User
+from ..module_utils.role import Role
+
+
+def get_role_uuids(module, rest_client: RestClient):
+    # CachedRestClient is beneficial here since we use for loop and make get requests to the same endpoint many times
```

Comment:
Oh no :) The CachedRestClient is intended to cache GET responses. It does not know when to flush cached data - so you can easily get outdated data from it. Until this is cache flush is implemented, it cannot be used in `wait until object.get_state()=="something"` . Please do not use it action modules.

If your code does "read user, modify user, read user", and REST calls are "GET users/uuid1, PUT users/uuid1, GET users/uuid1", the second GET will return data from first GET. Unless you manually flush cache by creating a new instance of CachedRestClient?

Or your code does not do second GET, as you can just assume new server state is same to PUT payload.

Eventually we can add better cache options:
- flush after each PUT/POST/DELETE
- flush manually
- flush after entry expires, based on time.

Reply:
Or maybe we do keep it here? It is not problem you are using CachedRestClient, problem is Jure and me didn't make it good enough. Integration tests do ensure this module DOES work. And if we can spare unneeded GET request, without making code hard to read - this is a good thing.

Just keep using it. Module does work. CachedRestClient can be made better too.

Just FYI. CC @domendobnikar  @anazobec

Reply:
@justinc1 There is no case in the module where there is "get -> post/patch -> get". I understand that in this case we would read old values. And in the playbook new instance of the client is generated every time we use the module so that's why the module does work :)
I agree the CachedRestClient could be updated in a way that only inside certain methods in module (in my case get_role_uuids) CachedRestClient would be created from RestClient - but in this case type(self.client) changes from Client to RestClient and and this is no good :) Or I could updated module so that "run" function accepts Client and then in each function inside the module I create RestClient or CachedRestClient - depends on what is needed. 
But since the module is working properly I can also leave it as it is.